gpio: add simulated gpio driver

cfriedt opened the pull request:
This change has similar goals to #26003 except that this is targeting gpio rather than i2c.

By simulating a GPIO controller and a number of pins, integration testing can be made simpler where high-level input needs to be validated against low-level I/O.

Fixes #26477

cfriedt commented:
Current status:
```console
*** Booting Zephyr OS build zephyr-v2.3.0-630-gcae423d86622  ***
Running test suite gpio_basic_test
===================================================================
START - test_gpio_port
Validate device GPIO_0
Check GPIO_0 output 0 connected to input 1
OUT 0 to IN 1 linkage works
- bits_physical
- pin_physical
- check_raw_output_levels
- check_logic_output_levels
- check_input_levels
- bits_logical
- check_pulls

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:510: check_pulls: (raw_in() not equal to false)
physical pull-down does not read low
 FAIL - test_gpio_port
===================================================================
START - test_gpio_callback_add_remove

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_callback_manage.c:104: test_callback_add_remove: (rc not equal to 0)
init_callback failed
 FAIL - test_gpio_callback_add_remove
===================================================================
START - test_gpio_callback_self_remove

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_callback_manage.c:153: test_callback_self_remove: (rc not equal to 0)
init_callback failed
 FAIL - test_gpio_callback_self_remove
===================================================================
START - test_gpio_callback_enable_disable

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_callback_manage.c:205: test_callback_enable_disable: (rc not equal to 0)
init_callback failed
 FAIL - test_gpio_callback_enable_disable
===================================================================
START - test_gpio_callback_variants
FAIL - test_callback@89. config PIN_IN interrupt fail: -38

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_callback_trigger.c:132: test_gpio_callback_variants: (test_callback(GPIO_INT_EDGE_FALLING) not equal to TC_PASS)
falling edge failed
 FAIL - test_gpio_callback_variants
===================================================================
Test suite gpio_basic_test failed.
===================================================================
PROJECT EXECUTION FAILED
FAILED: zephyr/CMakeFiles/run 
cd /home/chrisfriedt/workspace/zephyrproject/zephyr/build && /home/chrisfriedt/workspace/zephyrproject/zephyr/build/zephyr/zephyr.exe
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/local/bin/cmake --build /home/chrisfriedt/workspace/zephyrproject/zephyr/build --target run
```

cfriedt commented:
Need to add a mechanism to the backend for clearing interrupts still.

cfriedt commented:
This PR is close-ish to being done, so it makes sense to get some comments.
On my TODO list:
- [ ]  fill in parts of code marked TODO (e.g. checking pin / mask validity)
- [ ] add "backend" mechanisum for unpending an interrupt (i.e. marking it complete)
- [ ] document `gpio_sim.h`

sjg20 commented on drivers/gpio/Kconfig.sim:
```diff
@@ -0,0 +1,30 @@
+	bool "Simulated GPIO driver"
+	default n
+	help
+	  Enable simulated GPIO driver.
```
Can you add more details about what this does?

sjg20 commented on drivers/gpio/Kconfig.sim:
```diff
@@ -0,0 +1,30 @@
+
+if GPIO_SIM
+
+config GPIO_SIM_NPORTS
```
Seems like both of these configs could be in the device tree?

cfriedt replied:
Interesting - what would be your desired property name in that case?

cfriedt replied:
Done. Used the common ngpios DT binding.

sjg20 commented on include/drivers/gpio/gpio_sim.h:
```diff
@@ -0,0 +1,40 @@
+
+static inline int gpio_sim_output_get(struct device *port, gpio_pin_t pin)
+{
+	int r;
```
How about 'ret' which avoids a single-char var name and seems to be most common in Zephyr.

sjg20 commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,563 @@
+ * GPIO config / data
+ */
+
+struct gpio_sim_config {
```
Can you comment the struct?

cfriedt replied:
Done

sjg20 commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,563 @@
+	size_t i;
+	gpio_port_pins_t mask;
+
+	for (mask = 0, i = 0; i < CONFIG_GPIO_SIM_NPINS_PER_PORT; ++i) {
```
Instead of the config, can you put the value in gpio_sim_data once (hopefully from DT) and use it from there?

cfriedt replied:
Good suggestion

cfriedt replied:
Done

sjg20 commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,563 @@
+
+	for (mask = 0, i = 0; i < CONFIG_GPIO_SIM_NPINS_PER_PORT; ++i) {
+		if (config->flags[i] & flags) {
+			mask |= (1 << i);
```
You can use BIT()

sjg20 commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,563 @@
+int gpio_sim_input_set_masked(struct device *port, gpio_port_pins_t mask,
+			      gpio_port_value_t values)
+{
+	int r;
```
ret

sjg20 commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,563 @@
+				interrupts |= (1 << i);
+			}
+			break;
+
```
Are we supposed to have blank lines here?

cfriedt replied:
None of the linters seem to have complained. It was formatted with clang-format using Zephyr's .clang-format file.

cfriedt replied:
Removed the blank lines

sjg20 commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,563 @@
+	}
+
+	output_mask = gpio_sim_output_mask(config);
+
```
In general it feels like there are two many blank lines. This line seems linked to the if() below, so I think it should be together, for example

cfriedt replied:
Fixed.

sjg20 commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,563 @@
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+
+	if (0 == ((1 << pin) & config->common.port_pin_mask)) {
```
0 == ((1 << pin)

This seems to be a Zephyr style. The avoidance of ! is really surprising to me. I would write this as

unsigned int mask = BIT(pin);

if (!(config->common.port_pin_mask & mask))

but that seems to not be done in Zephyr? I do think a mask var would help in this function though

pabigot replied:
There's some push towards using MISRA compliant code in Zephyr, and MISRA has a concept of essential types which disallows implicit conversion between integer and boolean values.  Thus you'll often see `x == 0` or `x != 0` in cases where `x` is sufficient from a C language perspective.

Also while I happen to like Yoda comparisons they're not acceptable to `checkpatch`, so it'd be better to avoid `0 == x`.

The above expression should be:

    if ((config->common.port_pin_mask & BIT(pin)) == 0) {

sjg20 commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,563 @@
+	config->vals &= ~mask;
+	config->vals |= values;
+
+	r = 0;
```
I guess this function will be updated for error handling.

sjg20 commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,563 @@
+	return gpio_sim_init(dev, 2);
+}
+
+DEVICE_AND_API_INIT(gpio_sim_2, DT_INST_LABEL(2), gpio_sim_init_2,
```
If this moves to the DT (i.e. each bank is a separate node) then I hope this code can be simplified.

cfriedt replied:
Fixed.

aescolar commented on tests/drivers/gpio/gpio_basic_api/src/main.c:
```diff
@@ -115,6 +181,12 @@ static void board_setup(void)
 
 	pinmux_pin_set(pmx, PIN_OUT, PORT_PCR_MUX(kPORT_MuxAsGpio));
 	pinmux_pin_set(pmx, PIN_IN, PORT_PCR_MUX(kPORT_MuxAsGpio));
+#elif defined(CONFIG_BOARD_NATIVE_POSIX)
```
same for this ifdef

aescolar commented on tests/drivers/gpio/gpio_basic_api/boards/sim.conf:
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_GPIO_SIM=y
```
If I understand what you want you are trying to do*1: This file should be named tests/drivers/gpio/gpio_basic_api/boards/native_posix.conf

*1: Are you are trying to enable this config when native_posix is the target?

aescolar commented on tests/drivers/gpio/gpio_basic_api/boards/sim.overlay:
```diff
@@ -0,0 +1,13 @@
+/*
```
And this one would be named tests/drivers/gpio/gpio_basic_api/boards/native_posix.overlay

aescolar commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,563 @@
+#include <drivers/gpio.h>
+#include <drivers/gpio/gpio_sim.h>
+#include <errno.h>
+#include <stdio.h>
```
you are not using this header(?) => do not include it

cfriedt replied:
I had used snprintf at some point. It's no longer needed.

cfriedt commented on tests/drivers/gpio/gpio_basic_api/src/main.c:
```diff
@@ -20,6 +20,72 @@
 #include "soc.h"
 #endif
 
+#if defined(CONFIG_BOARD_NATIVE_POSIX)
```
Fixed.

cfriedt committed:
gpio: add driver for emulated GPIO

The emulated GPIO controller will aid in automated
integration testing.

Fixes #26477

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
Not sure how to reference a source file in Doxygen .. would be nice to point to exmaples with hyperlinks for the doxygen in `gpio_sim.h`.

Also, any pointers for un-pending interrupts would be appreciated.

cfriedt commented:
buildkite is failing because of some unrelated test in `qemu_arc_hs`. Hopefully that gets fixed soon.

galak commented on dts/bindings/gpio/zephyr,sim-gpio.yaml:
```diff
@@ -0,0 +1,22 @@
+# Copyright 2020, Friedt Professional Engineering Services, Inc
+# SPDX-License-Identifier: Apache-2.0
+
+description: GPIO Simulator
```
Can you add a little more detail to the description.  Something like 'Native Posix simulated GPIO - See drivers/gpio/Kconfig.sim... for more details'

nashif replied:
>  Something like 'Native Posix simulated GPIO

I thought it was agreed this is not specific to Native Posix...

aescolar replied:
It is not :)

cfriedt commented:
I still need to add at least one test for clearing interrupt status. In Zephyr GPIO drivers, is the convention that the IRQ clears the interrupt status bits before calling callbacks?

carlescufi commented:
@mnkp and @pabigot could you please take a look?

pabigot commented on tests/drivers/gpio/gpio_basic_api/src/main.c:
```diff
@@ -20,6 +20,68 @@
 #include "soc.h"
 #endif
 
+#if defined(CONFIG_GPIO_SIM)
```
Please put all this in a separate source file that's added to `target_sources` only when this option is enabled.

pabigot commented on tests/drivers/gpio/gpio_basic_api/src/main.c:
```diff
@@ -115,6 +177,12 @@ static void board_setup(void)
 
 	pinmux_pin_set(pmx, PIN_OUT, PORT_PCR_MUX(kPORT_MuxAsGpio));
 	pinmux_pin_set(pmx, PIN_IN, PORT_PCR_MUX(kPORT_MuxAsGpio));
+#elif defined(CONFIG_GPIO_SIM)
```
Use a local `extern struct gpio_callback *gpio_sim_callback;`  declaration to expose the content from the separate source file.

pabigot reviewed (changes requested):
This looks interesting; please review comments below.  I think all are requested but if some seem inappropriate let me know.

cfriedt commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,630 @@
+	uint32_t *const flags;
+	/** Number of pins available in the given GPIO controller instance */
+	const gpio_pin_t num_pins;
+	/** Input or output values for each pin */
```
Nice - yeah, I wasn't sure which struct contained modifiable bits. Thanks.

cfriedt commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,630 @@
+LOG_MODULE_REGISTER(gpio_sim);
+
+#define GPIO_SIM_INT_BITMASK							\
+	(GPIO_INT_DISABLE | GPIO_INT_DEBOUNCE | GPIO_INT_ENABLE |		\
```
It's really just used to clear a bunch of bits all at once.

pabigot replied:
Yes; and one of those bits doesn't belong in the set.

cfriedt replied:
Done! Thanks for catching that :-)

pabigot commented:
Didn't notice these before:

> * how to reference a source file in Doxygen (would be nice to point to examples with hyperlinks for the doxygen in `gpio_sim.h`)

I don't know.

> * un-pending interrupts

That would normally be done in the ISR, which here is simulated by `gpio_sim_pend_interrupt()` being invoked when pins change via `gpio_sim_input_set_masked()`.

Normally all interrupts that are set in the call to `gpio_fire_callbacks()` would be cleared before it's invoked.   So there is no need to store them in the `interrupts` data word, unless you want to support an interrupt architecture where the condition can be captured but doesn't trigger as an interrupt until the pin is configured to interrupt mode.

What you do need to do, though, is fix the pend code for level interrupts.  One way to do that could be: after calling `gpio_fire_callbacks()` go through and check the level interrupts again, and if any are pended re-invoke `gpio_fire_callbacks()`.  Repeat until no pins are found to be pended.  This would of course deadlock if the callbacks don't reconfigure the interrupts or change the state that triggered the level---which is exactly how real hardware behaves.

> * are `__syscall` prefixes required for the `gpio_sim.h` API calls?

Only if you expect the simulator manipulation functions to be invoked from a user process.  I would guess that's unlikely.

cfriedt commented:
Thanks for all of the suggestions. I'm in a rabbit hole elsewhere but will get to this soon!

cfriedt commented:
> Didn't notice these before:
> > * un-pending interrupts
> 
> That would normally be done in the ISR, which here is simulated by `gpio_sim_pend_interrupt()` being invoked when pins change via `gpio_sim_input_set_masked()`.
> 
> Normally all interrupts that are set in the call to `gpio_fire_callbacks()` would be cleared before it's invoked. So there is no need to store them in the `interrupts` data word, unless you want to support an interrupt architecture where the condition can be captured but doesn't trigger as an interrupt until the pin is configured to interrupt mode.
> 
> What you do need to do, though, is fix the pend code for level interrupts. One way to do that could be: after calling `gpio_fire_callbacks()` go through and check the level interrupts again, and if any are pended re-invoke `gpio_fire_callbacks()`. Repeat until no pins are found to be pended. This would of course loop forever if the callbacks don't reconfigure the interrupts or change the state that triggered the level---which is exactly how real hardware behaves.

Thanks @pabigot

I'll have a look at unpending this afternoon, and I think I addressed all of your other points as well.

cfriedt commented:
> > Didn't notice these before:
> > What you do need to do, though, is fix the pend code for level interrupts. One way to do that could be: after calling `gpio_fire_callbacks()` go through and check the level interrupts again, and if any are pended re-invoke `gpio_fire_callbacks()`. Repeat until no pins are found to be pended. This would of course loop forever if the callbacks don't reconfigure the interrupts or change the state that triggered the level---which is exactly how real hardware behaves.

@pabigot - please have a look at `gpio_sim_pend_interrupt()` again. I believe this should take into account level interrupt behaviour.

I could put together another test for that though, if you like. Interrupt bits are generated in a separate function now. The first time, it will generate interrupt bits for both edge and level interrupts. On the second and subsequent checks, only level-interrupt bits are generated. Level interrupts continue to be generated until the level is changed.

pabigot reviewed (dismissed):
OK, this got long, so I'm stopping now.  Not everything here is "change requested", but a lot is.

At a high level I'm concerned about simulator fidelity and validation.  That in-tree GPIO tests pass with this is insufficient: those tests were written based on an assumption that hardware GPIO peripherals behave correctly if the configuration is correct.

The simulator should have its own dedicated test, preferably one run with coverage enabled, that checks things like whether changing GPIO configuration away from and back to a particular mode inadvertently clears state that would be persisted in real hardware.

Examples that I don't think this does quite right include the input/output/hi-Z state of a pin; handling pull configurations; and whether an interrupt event captured by the hardware but ignored due to interrupts being enabled is triggered immediately if the pin configuration is changed.

I would really like to see a documented shared understanding of the role simulation will play in Zephyr evolution, so we don't have conflicting models.  This one seems to have no external dependencies, and could be run on real microcontroller hardware (e.g. to simulate an external device that isn't available in hardware).  The one originally proposed in #26003 before it got derailed might be able to do more by being "limited" to a single platform that's pretty powerful.  Or maybe they're entirely compatible and will work together.

If we don't coordinate things, that's less likely to be true.

cfriedt commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,639 @@
+	/** Pointer to an array of flags is @a num_pins in size */
+	uint32_t *const flags;
+	/** Input or output values for each pin */
+	gpio_port_value_t vals;
```
Sure. Yeah, I guess it's possible to write to the output register of a GPIO even though it's configured for input. My code currently assumes pull-down resistors, but you're right in that I should also consider pull-up as well as hi-z.

cfriedt commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,639 @@
+ * The caller must ensure that @ref gpio_sim_data.mu is locked.
+ *
+ * @param port The simulated GPIO device pointer
+ * @param flags The flags to match
```
Ah ok. The design is "has exactly these flags". I can update the docs to be more clear about that.

cfriedt commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,639 @@
+	prev_values = drv_data->vals;
+
+	output_mask = gpio_sim_output_mask(port);
+	/* silently correct mask (otherwise CI fails) */
```
Sure. I can do that, if that's the convention with Zephyr.

cfriedt commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,639 @@
+		switch (trig) {
+		case GPIO_INT_TRIG_LOW:
+		case GPIO_INT_TRIG_HIGH:
+		case GPIO_INT_TRIG_BOTH:
```
Interesting. Are you thinking that would be a good option to expose through DT bindings?

pabigot replied:
Yes, properties specific to this driver could allow per-instance control of the features and behavior it provides.

cfriedt replied:
@pabigot - added interrupt capabilities. Please have a look.

pabigot commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,639 @@
+	drv_data->vals |= values;
+	values = drv_data->vals;
+
+	gpio_sim_pend_interrupt(port, mask, prev_values, values);
```
Actually I just remembered (perhaps incorrectly) that Zephyr's `k_mutex` is recursive, in which case this could work fine.

It's still pretty unusual to use `k_mutex` instead of `k_sem` for driver serial access, though I'm trying to get buy-in to change that.

cfriedt commented on boards/posix/native_posix/native_posix.dts:
```diff
@@ -6,6 +6,7 @@
 
 /dts-v1/;
 #include <posix/posix.dtsi>
+#include <dt-bindings/gpio/gpio.h>
```
> OK, this got long, so I'm stopping now. Not everything here is "change requested", but a lot is.

No problem - I'll make the requested changes when I have some time. I hope you don't mind that I keep the PR open for a while.

> At a high level I'm concerned about simulator fidelity and validation. That in-tree GPIO tests pass with this is insufficient: those tests were written based on an assumption that hardware GPIO peripherals behave correctly if the configuration is correct.

Yeah, that was mainly intended to keep things simple at first. E.g. the assumption was made that inputs have pull-down resistors, but that's obviously not always the case. Similarly, went with the idea of GPIO registers holding values for input or output, but not both, but that's not the case either in a lot of hardware. Do you think it's better to use a `gpio_port_pins_t` for input, output & hi-z? I'm just wondering because maybe I can store all pin attributes in a 32-bit value (like I do for most of the GPIO configuration). 

Speaking of GPIO configuration, I think it would be pretty useful to add an API call for querying that on a per-pin basis - kind of the inverse of `gpio_config()`. It would be extremely useful in my application, and as you can see, I needed to expose that functionality just to get GPIO simulation to work with "wiring".

> The simulator should have its own dedicated test, preferably one run with coverage enabled, that checks things like whether changing GPIO configuration away from and back to a particular mode inadvertently clears state that would be persisted in real hardware.

Sure - how do I go about enabling coverage? Can you point me to an example? If you're more specific about the state that should be persisted, I can add tests for that.

> Examples that I don't think this does quite right include the input/output/hi-Z state of a pin; handling pull configurations; and whether an interrupt event captured by the hardware but ignored due to interrupts being enabled is triggered immediately if the pin configuration is changed.

Agreed. I'll look at in/out/hi-z. 

I might need you to go into specifics for the latter part of that though. No rush, just whenever you have some time.

> I would really like to see a documented shared understanding of the role simulation will play in Zephyr evolution, so we don't have conflicting models. This one seems to have no external dependencies, and could be run on real microcontroller hardware (e.g. to simulate an external device that isn't available in hardware). The one originally proposed in #26003 before it got derailed might be able to do more by being "limited" to a single platform that's pretty powerful. Or maybe they're entirely compatible and will work together.
> 
> If we don't coordinate things, that's less likely to be true.

Sure. My main purpose is to be able to "mock out" a GPIO controller so that I can create integration and system tests that can run in CI. In the end, we're just dealing with information, and I'm particularly focused on the communication of that information at the moment, and not whether the "mock" (model) is perfect. So the device-under-test in my case is a communications API that is designed specifically for dealing with embedded I/O. However, I think it's important to make that model perfect (or as close to perfect as can be expected) in the long run. Certainly, at some point, there will be a different framework that will interact with I/O and it would most likely benefit by being able to mock out I/O.

If there is a different model that this could be useful for, I'd be happy to hear about that as well.

sjg20 commented on drivers/gpio/Kconfig.sim:
```diff
@@ -0,0 +1,17 @@
+	  Enable the simulated GPIO driver. Mainly used for testing, this
+	  driver allows for an arbitrary number of simulated GPIO controllers
+	  to be instantiated. Furthermore, the simulated pins can be "wired"
+	  using using the regular GPIO callback API and the additional API
```
s/using //

sjg20 commented on drivers/gpio/Kconfig.sim:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config GPIO_SIM
+	bool "Simulated GPIO driver"
```
I suppose we need to decide whether to use simulator or emulator for these features.

I feel that emulator is a slightly better term, since we are not actually trying to model the internal workings of the device, but just make it appear to work in the same way from the outside.

If we used simulators, we would be trying to model the internal workings of each device, e.g. a SPICE simulation of the workings of a PMIC.

So I think emulator makes more sense.

pabigot replied:
From https://stackoverflow.com/a/1584701/236848:

> Emulation is the process of mimicking the outwardly observable behavior to match an existing target. The internal state of the emulation mechanism does not have to accurately reflect the internal state of the target which it is emulating.
>
> Simulation, on the other hand, involves modeling the underlying state of the target. The end result of a good simulation is that the simulation model will emulate the target which it is simulating.

I agree, emulation is the correct term here, and I will adjust my own usage.  A goal from my perspective is to be able to synthesize behavior corresponding to faults to ensure they're detected and handled properly.

pabigot replied:
Except maybe not.... This driver *is* intended to model the internal workings of a GPIO peripheral, in that it maintains per-pin state and configuration and generates interrupts where a real device would.

For a temperature sensor code that simply records a specific temperature provided by a test framework and generates I2C responses that match what would be generated by a real sensor could be an emulator.  If the code instead implemented an algorithm to automatically adjust the temperature with the passage of time, generating threshold events as it changes, might be better described as a simulator.

I suppose it depends on the fidelity of "internal state", and more importantly on whether internal *behavior* is also exposed.

More discussion is needed, probably in the context of #27531.

sjg20 replied:
Yes it depends. You could argue that a real simulation of the GPIO would include voltage levels and model its reaction to the the drain and pull-ups, etc.

But since it doesn't simulate any particular GPIO chip or SoC GPIO block, and it doesn't really make sense to simulate a 'generic' GPIO, perhaps emulator is better?

cfriedt replied:
Going with "emulate", since that seems to already be the decision made in upstream.

sjg20 commented on include/drivers/gpio/gpio_sim.h:
```diff
@@ -0,0 +1,132 @@
+ * @ingroup gpio_interface
+ * @{
+ *
+ * Behaviour of virtual GPIO is application-defined. As-such, each
```
Here we are introducing another term, 'virtual'. I suggest 'emulated GPIO'.

sjg20 commented on include/drivers/gpio/gpio_sim.h:
```diff
@@ -0,0 +1,132 @@
+ *
+ * @param port The simulated GPIO port
+ * @param pin The pin to modify
+ * @param value New values to assign to @p pin
```
Could you mention polarity here? I assume this is active high always

pabigot replied:
The simulator should be operating on physical (line) levels.  Logical level translation is handled in the common API wrappers.

But yes, the API for manipulating the GPIO state external to the standard GPIO API should be clear that it uses only physical level values.  Unless we want a replicated API that works with logic level values.

cfriedt replied:
Considering this resolved.

sjg20 commented on tests/drivers/gpio/gpio_basic_api/boards/native_posix.overlay:
```diff
@@ -0,0 +1,17 @@
+
+/ {
+	resources {
+		compatible = "test,gpio_basic_api";
```
Compatible strings should use - not _

cfriedt replied:
Created #29689. That's a change that should be applied to the actual dts binding as well as all existing boards.

sjg20 commented:
Any progress on this one?

cfriedt commented:
> Any progress on this one?

I have not had an iota of time to revisit it. 

Customer requires

* dynamic pthread stack allocation
* dns-sd
* various RF modifications for the cc1352r

Reworking these changes is not funded per se, so it has to take the bench until I have some time. I currently have a workaround for the first one, but neither of rhe latter two.

Just about done dns-sd.

cfriedt commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,639 @@
+ * @param flags The flags to match
+ */
+static gpio_port_pins_t
+gpio_sim_get_mask_of_pins_with_flags(struct device *port, uint32_t flags)
```
I'm fine with a shorter name.

cfriedt commented on drivers/gpio/gpio_sim.c:
```diff
@@ -0,0 +1,639 @@
+{
+	size_t i;
+	gpio_port_pins_t mask;
+	struct gpio_sim_data *const drv_data =
```
Yeah, I guess it's a bit of a habit. Not explicitly necessary. I can remove it if it's confusing.

cfriedt committed:
tests: gpio_basic_api: support for emulated GPIO

This adds support for emmulated GPIO (CONFIG_GPIO_EMUL=y) and
additionally allows BOARD=native_posix and
BOARD=native_posix_64 to run the gpio_basic_api test suite.

Fixes #26477

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

sjg20 reviewed (commented):
This looks good to me. The test passes. I haven't tried to press it into service but I expect to do that after this goes in.

What is the 'stashing' commit for?

cfriedt commented:
> What is the 'stashing' commit for?

An additional goal right now is to ensure that tests also pass for the `gpio_api_1pin` suite as well as checking off the box below:
```
- [ ] configurable interrupt mode via DT bindings - level, edge, and / or double-edge
```
Incidentally, is G using Z for a new BMC? Sounds very interesting.. If only I got > 4/5 stars when I intereviewed!

cfriedt committed:
tests: gpio_api_1pin: support for emulated GPIO

This adds support for emulated GPIO (CONFIG_GPIO_EMUL=y) and
additionally allows BOARD=native_posix and
BOARD=native_posix_64 to run the gpio_api_1pin test suite.

Fixes #26477

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
The `gpio_basic_api` and `gpio_api_1pin` test suites pass now but I feel like it's a bit hackish.

@pabigot - any thoughts?

The `gpio_api_1pin` test suite ensures that callbacks are called maximally once per event.

The last modification I made was to not duplicate callbacks when both an interrupt fires (the usual behaviour) and when an output pin changes value (for emulation of 'wiring'). I was getting two callbacks for a single pin change. I guess that I could add another variable saying which pins have had interrupts fire, and then mask off pins that were already handled by their input interrupts?

sjg20 commented on tests/drivers/gpio/gpio_basic_api/src/gpio_emul_callback.c:
```diff
@@ -0,0 +1,74 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifdef CONFIG_GPIO_EMUL
```
Can this move to the CMakeLists.txt file?

cfriedt replied:
I think that should work - let me give it a shot.

cfriedt replied:
Looks like it won't due to globbing.
```
FILE(GLOB app_sources src/*.c)
```

pabigot replied:
Change `CMakeLists.txt`:

```
FILE(GLOB app_sources src/main.c src/test*.c)
target_sources(app PRIVATE ${app_sources})
target_sources_ifdef(CONFIG_GPIO_EMUL app PRIVATE src/gpio_emul_callback.c)
```
and get rid of the whole-file `#if/#endif`.

sjg20 commented on tests/drivers/gpio/gpio_basic_api/src/gpio_emul_callback.c:
```diff
@@ -0,0 +1,74 @@
+ * PIN_IN. When PIN_OUT is set high or low, PIN_IN must be set
+ * correspondingly, as if a wire were connecting the two.
+ */
+
```
drop blank line?

cfriedt replied:
Sure

sjg20 reviewed (commented):
This looks good to me. One question and one nit

pabigot commented on drivers/gpio/gpio_emul.c:
```diff
@@ -0,0 +1,596 @@
+	/** Interrupt status for each pin */
+	gpio_port_pins_t interrupts;
+	/** Mutex to synchronize accesses to driver data and config */
+	struct k_mutex mu;
```
I think we're generally using a spinlock for this.  Mutex has different priority handling behavior and allows recursion, IIRC.

cfriedt replied:
I need it for reentrance.

pabigot replied:
OK, that makes sense: e.g. allowing reconfiguration from within an interrupt callback.

pabigot commented on drivers/gpio/gpio_emul.c:
```diff
@@ -0,0 +1,596 @@
+	uint32_t flags)
+{
+	size_t i;
+	gpio_port_pins_t matched;
```
I'd rather see `matched` initialized here instead of the entry to the for loop.  It just looks weird there.

pabigot commented on drivers/gpio/gpio_emul.c:
```diff
@@ -0,0 +1,596 @@
+ * @return a mask of the pins with matching @p flags
+ */
+static gpio_port_pins_t
+get_pins_with_flags(const struct device *port, uint32_t mask,
```
Shouldn't `mask` and `flags` be `gpio_port_value_t`?

cfriedt replied:
`gpio_port_pin_t` and `gpio_flags_t` - good catch!

pabigot commented on drivers/gpio/gpio_emul.c:
```diff
@@ -0,0 +1,596 @@
+	/** Common @ref gpio_driver_data */
+	struct gpio_driver_data common;
+	/** Pointer to an array of flags is @a num_pins in size */
+	uint32_t *flags;
```
Are these supposed to be configuration flags like `GPIO_INPUT`, something else, or both?  If the former it should be `gpio_port_value_t`, right?  I see this being used with configuration flags, it really shouldn't store anything else.

cfriedt replied:
`gpio_flags_t` - thanks for catching that :-)

pabigot commented on drivers/gpio/gpio_emul.c:
```diff
@@ -0,0 +1,596 @@
+	     interrupts; gpio_emul_gen_interrupt_bits(
+		     port, mask, prev_values, values, &interrupts, false)) {
+		gpio_fire_callbacks(&drv_data->callbacks, port, interrupts);
+	}
```
My head hurts.  Maybe:
```
	gpio_emul_gen_interrupt_bits(port, mask, prev_values, values,
				     &interrupts, true);
	while (interrupts != 0) {
		gpio_fire_callbacks(&drv_data->callbacks, port, interrupts);
		gpio_emul_gen_interrupt_bits(port, mask, prev_values, values,
					     &interrupts, false);
	}
```
Please?

pabigot commented on drivers/gpio/Kconfig.emul:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config GPIO_EMUL
+	bool "Emulated GPIO driver"
```
Please add ` [EXPERIMENTAL]` to the option title at least for 2.5.  We should get more users and experience with it before deciding whether the driver-specific API is what we're going to want.

pabigot reviewed (dismissed):
Suggestion to move to `k_semaphore` for mutex because that's more common, but since we're not consistent in Zephyr I won't insist.

I think a lot of `uint32_t`s should be something more specific like `gpio_port_value_t`.  That should be cleaned up, if so.

Also the `for` statement where entry and exit statements are longer than the body really ought to be changed.

I think I had some visions of being able to use properties the binding to control implicit behavior on a per-instance basis, but I don't remember the details so I'm OK with having this go in.  I do want it flagged as experimental so we don't have to fight with users if the API needs to change.

Otherwise I think this is close.

cfriedt commented:
> Suggestion to move to `k_semaphore` for mutex because that's more common, but since we're not consistent in Zephyr I won't insist.

I think the mutex is necessary in one or 2 locations due to reentrancy.

> I think a lot of `uint32_t`s should be something more specific like `gpio_port_value_t`.  That should be cleaned up, if so.

Ok, will have a look.

> Also the `for` statement where entry and exit statements are longer than the body really ought to be changed.

Hehe.. fair enough - will look at that today.

> I think I had some visions of being able to use properties the binding to control implicit behavior on a per-instance basis, but I don't remember the details so I'm OK with having this go in.  I do want it flagged as experimental so we don't have to fight with users if the API needs to change.

Yep, those features were added. I think there is a separate commit for it, including bindings. It makes support for rising / falling / dual-edge as well as high / low / dual-level optional on a per-instance basis.

I broke it all out into individual bools in DT and the tests / samples enable all interrupt types Have not yet written test cases / overlays for every permutation though.

pabigot commented:
> > Suggestion to move to `k_semaphore` for mutex because that's more common, but since we're not consistent in Zephyr I won't insist.
> 
> I think the mutex is necessary in one or 2 locations due to reentrancy.

I'm not saying you don't need a mutex, but the mutex type you chose isn't what we normally use.  [k_mutex](https://docs.zephyrproject.org/latest/reference/kernel/synchronization/mutexes.html) is reentrant and priority inheriting; most things use [k_sem](https://docs.zephyrproject.org/latest/reference/kernel/synchronization/semaphores.html#suggested-uses) with a maximum count of 1.

pabigot commented on drivers/gpio/gpio_emul.c:
```diff
@@ -0,0 +1,687 @@
+ * @return a mask of the pins with matching @p flags
+ */
+static gpio_port_pins_t
+get_pins_with_flags(const struct device *port, gpio_port_pins_t mask,
```
`gpio_flags_t` mask, since the input mask is also flags of interest, not pins of interest.

cfriedt replied:
Yes - thanks!

cfriedt commented:
Uh-oh - will need to back off my last changeset.

cfriedt commented:
I think I saw a issue with a separate driver failing some tests, so rebased. Hopefully it's all green again.

pabigot commented on dts/bindings/gpio/zephyr,gpio-emul.yaml:
```diff
@@ -0,0 +1,40 @@
+    label:
+      required: true
+
+    rising-edge:
```
Sorry, I missed this.  Please provide at least basic documentation on the properties, e.g.:

    description: Enables support for rising edge interrupt detection

pabigot commented on dts/bindings/gpio/zephyr,gpio-emul.yaml:
```diff
@@ -0,0 +1,40 @@
+    low-level:
+      type: boolean
+
+    dual-level:
```
Can't the `duel` properties be defined by having both options (levels or edges) enabled?

cfriedt replied:
Not supporting dual but supporting rising / falling or high / low would clearly be very bad hardware design, but for flexibility, I opted to keep them distinct.

pabigot replied:
Maybe I was unclear.  What would it mean to specify that both rising and falling edge detection are supported, but that both edge detection is not supported?  Or that both edge detection is supported, but (say) falling edge detection is not?

By having a distinct property to indicate dual support you open the possibility of inconsistent configurations.  Or am I missing something?

cfriedt replied:
Yep, there is the distinct possibility of having inconsistent configurations.

E.g. both rising and falling edge might be unsupported, but dual edge is supported, not that you would necessarily want to do that.

It's flexible enough to support interrupt configuration you can think of, probably even no interrupt capabilities.

pabigot replied:
I don't usually say this because I like flexibility, but allowing support for dual edge/level without allowing both elemental single edge/levels is a level of flexibility I don't think we need: "inconsistent" isn't "flexible".

Please remove the `dual-*` properties.  That's enough to support no interrupts, one sided, other sided, and both sided which together cover the meaningful configurations.

cfriedt replied:
OK

pabigot commented on drivers/gpio/gpio_emul.c:
```diff
@@ -0,0 +1,687 @@
+	return gpio_manage_callback(&drv_data->callbacks, cb, set);
+}
+
+static uint32_t gpio_emul_get_pending_int(const struct device *dev)
```
s/uint32_t/gpio_port_pins_t/

pabigot replied:
Also in `gpio_emul_flags_get()` and related API (apparently that comment didn't get saved in the review).

pabigot reviewed (dismissed):
Missed looking at the binding, found a couple more `uint32_t`.

cfriedt commented:
> Missed looking at the binding, found a couple more `uint32_t`.

I think I got the last of them. Thanks!

pabigot commented on drivers/gpio/gpio_emul.c:
```diff
@@ -0,0 +1,699 @@
+ * @param caps A bitmask of capabilities
+ */
+static inline bool gpio_emul_config_has_caps(const struct device *port,
+		enum gpio_emul_interrupt_cap caps) {
```
This is useful, but the `caps` argument needs to be type `int`, because this is invoked with arguments that are the bitwise or of two of the enum values, which is not itself an enum value.

cfriedt replied:
True, yea

cfriedt committed:
gpio: emul: support configurable interrupt capabilities

This change adds support for configurable interrupt capabilities
in the emulated GPIO controller via Devicetree bindings.

Fixes #26477

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>
